store/postgres/src/writable.rs

```diff
@@ -325,37 +351,477 @@ impl WritableStore {
+        block_ptr: BlockPtr,
+        firehose_cursor: Option<String>,
+    },
+}
+
+/// A queue that asynchronously writes requests queued with `push` to the
+/// underlying store and allows retrieving information that is a combination
+/// of queued changes and changes already committed to the store.
+struct Queue {
+    store: Arc<SyncStore>,
+    /// A queue of pending requests. New requests are appended at the back,
+    /// and popped off the front for processing. When the queue only
+    /// contains `Write` requests block numbers in the requests are
+    /// increasing going front-to-back. When `Revert` requests are queued,
+    /// that is not true anymore
+    queue: BoundedQueue<Arc<Request>>,
+
+    /// The write task puts errors from `transact_block_operations` here so
+    /// we can report them on the next call to transact block operations.
+    write_err: Mutex<Option<StoreError>>,
+
+    /// True if the background worker ever encountered an error. Once that
+    /// happens, no more changes will be written, and any attempt to write
+    /// or revert will result in an error
+    poisoned: AtomicBool,
 }
 
-#[allow(dead_code)]
-pub struct WritableAgent {
-    store: Arc<WritableStore>,
+impl Queue {
+    /// Create a new queue and spawn a task that processes write requests
+    fn start(
+        logger: Logger,
+        store: Arc<SyncStore>,
+        capacity: usize,
+        registry: Arc<dyn MetricsRegistry>,
+    ) -> Arc<Self> {
+        async fn start_writer(queue: Arc<Queue>, stopwatch: StopwatchMetrics) {
+            loop {
+                // We peek at the front of the queue, rather than pop it
+                // right away, so that query methods like `get` have access
+                // to the data while it is being written. If we popped here,
+                // these methods would not be able to see that data until
+                // the write transaction commits, causing them to return
+                // incorrect results.
+                let res = match queue.queue.peek().await.as_ref() {
+                    Request::Write {
+                        block_ptr: block_ptr_to,
+                        firehose_cursor,
+                        mods,
+                        data_sources,
+                        deterministic_errors,
+                    } => queue.store.transact_block_operations(
+                        block_ptr_to,
+                        firehose_cursor.as_deref(),
+                        mods,
+                        stopwatch.cheap_clone(),
+                        data_sources,
+                        deterministic_errors,
+                    ),
+                    Request::Revert {
+                        block_ptr,
+                        firehose_cursor,
+                    } => queue
+                        .store
+                        .revert_block_operations(block_ptr.clone(), firehose_cursor.as_deref()),
```

Comment:
Now that this is called from a non-blocking tokio task, `transact_block_operations` and  `revert_block_operations` should be asyncified.

Reply:
I tried asyncifying `transact_block_operations` for a bit, but it requires that `DeploymentStore.transact_block_operations` clone all its arguments to own them because they get put into a background task since Rust doesn't understand that the `with_conn(|_, _| ...).await` that uses the arguments has to return before `transact_block_operations` returns and wants references with a static lifetime.

Rather than copy large amounts of data, I think we should hold off on asyncifying.

Reply:
Then we should spawn the writer in a blocking task.

Reply:
I didn't put the whole writer into a blocking task since that blocks the process from exiting; rather, I run the actual store operations as a blocking task now.